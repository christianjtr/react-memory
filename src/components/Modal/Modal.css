.modal {
  @apply fixed z-10 overflow-y-auto top-0 w-full left-0;
}

.modal > .modal__contentainer-wrapper {
  @apply flex items-center justify-center pt-4 px-4 pb-20 text-center sm:block sm:p-0;
}

.modal .modal__center-anchor {
  @apply sm:inline-block align-middle h-screen;
}

.modal .modal__overlay {
  @apply fixed inset-0 transition-opacity;
}

.modal .modal__overlay > .modal__overlay-mask {
  @apply absolute inset-0 bg-gray-900 opacity-75;
}

.modal .modal__container {
  @apply inline-block origin-center bg-white rounded-lg text-left overflow-hidden shadow-xl transform transition-all sm:align-middle sm:max-w-lg w-1/2;
}

.modal .modal__body,
.modal__footer {
  @apply px-4;
}

.modal .modal__body {
  @apply bg-white pt-5 pb-4 sm:p-6 sm:pb-4;
}

.modal .button {
  @apply bg-blue-500 hover:bg-blue-700 text-white py-2 px-4 rounded;
}
