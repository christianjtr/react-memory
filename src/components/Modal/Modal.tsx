import React from 'react';
import './Modal.css';

import { useGameContext } from '../../hooks';

export interface ModalProps {
  onOk: () => void;
}

const Modal: React.FC<ModalProps> = (props: ModalProps): React.ReactElement | null => {
  const { onOk: handleOnOkCallBack } = props;

  const {
    state: { isGameOver, score },
  } = useGameContext();

  if (!isGameOver) return null;

  return (
    <div className="modal" id="modal">
      <div className="modal__contentainer-wrapper">
        <div className="modal__overlay">
          <div className="modal__overlay-mask" />
        </div>
        <span className="modal__center-anchor">&#8203;</span>
        <div
          className="modal__container"
          role="dialog"
          aria-modal="true"
          aria-labelledby="modal-headline">
          <div className="modal__body text-center text-2xl">
            <p>
              <span className="text-center text-5xl">
                GAME
                <br />
                OVER
              </span>
            </p>
            <p className="my-5">{`Score: ${score}`}</p>
            <button
              id="modal-button"
              className="button"
              title="New game!"
              aria-label="New game!"
              onClick={handleOnOkCallBack}>
              New game!
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Modal;
