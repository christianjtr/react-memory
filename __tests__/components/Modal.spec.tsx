import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, cleanup, act, fireEvent } from '@testing-library/react';
import { createRenderer } from 'react-test-renderer/shallow';

import Modal from '../../src/components/Modal/Modal';

jest.mock('../../src/hooks', () => ({
  useGameContext: jest.fn().mockReturnValue({
    state: {
      isGameOver: true,
      score: 100,
    },
  }),
}));

describe('Modal component', () => {
  const handleOnOkCallBack = jest.fn();

  afterEach(() => {
    cleanup();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should render the component correctly', async () => {
    const renderer = createRenderer();
    const component = renderer.render(<Modal onOk={handleOnOkCallBack} />);

    expect(component).toBeDefined();
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it('should check elements when the modal has been rendered', async () => {
    const { container } = render(<Modal onOk={handleOnOkCallBack} />);

    const modalElement = container.getElementsByClassName('modal');
    const scoreElement = screen.getByText('Score: 100');

    expect(modalElement.length).toBe(1);
    expect(scoreElement).toBeDefined();
  });

  it('should call handleOnOkCallBack when clicking on the button', async () => {
    render(<Modal onOk={handleOnOkCallBack} />);

    const buttonElement = screen.getByRole('button');

    expect(buttonElement).toBeDefined();
    expect(typeof buttonElement.onclick === 'function').toBeTruthy();

    act(() => {
      fireEvent.click(buttonElement);
    });

    expect(handleOnOkCallBack).toHaveBeenCalled();
  });

  it('should not display the modal if the game is not over', async () => {
    jest.doMock('../../src/hooks');
    const { useGameContext } = await import('../../src/hooks');
    (useGameContext as jest.Mock).mockReturnValue({
      state: {
        isGameOver: false,
        score: 100,
      },
    });

    const { container } = render(<Modal onOk={handleOnOkCallBack} />);
    const modalElement = container.getElementsByClassName('modal');

    expect(modalElement.length).toBe(0);
  });
});
