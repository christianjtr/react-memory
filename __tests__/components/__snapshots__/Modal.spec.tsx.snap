// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render the component correctly 1`] = `
<div
  className="modal"
  id="modal"
>
  <div
    className="modal__contentainer-wrapper"
  >
    <div
      className="modal__overlay"
    >
      <div
        className="modal__overlay-mask"
      />
    </div>
    <span
      className="modal__center-anchor"
    >
      â€‹
    </span>
    <div
      aria-labelledby="modal-headline"
      aria-modal="true"
      className="modal__container"
      role="dialog"
    >
      <div
        className="modal__body text-center text-2xl"
      >
        <p>
          <span
            className="text-center text-5xl"
          >
            GAME
            <br />
            OVER
          </span>
        </p>
        <p
          className="my-5"
        >
          Score: 100
        </p>
        <button
          aria-label="New game!"
          className="button"
          id="modal-button"
          onClick={[MockFunction]}
          title="New game!"
        >
          New game!
        </button>
      </div>
    </div>
  </div>
</div>
`;
